### PR TITLE
Fix gossip deadlock

### PR DESCRIPTION
**Pull request by tamird**

### CODE REVIEW

File: gossip/gossip.go [hunk 1]
```diff
@@ -454,61 +454,81 @@ func (g *Gossip) manage(stopper *util.Stopper) {
...
-			// and there are still unused bootstrap hosts, signal bootstrapper
-			// to try another.
-			if g.outgoing.len()+g.incoming.len() == 0 {
-				g.stalled <- struct{}{}
-			} else if g.is.getInfo(KeySentinel) == nil {
-				log.Warningf("missing sentinel gossip %s; assuming partition and reconnecting", g.is.NodeID)
-				g.stalled <- struct{}{}
+func (g *Gossip) doCheckTimeout(stopper *util.Stopper) {
+	g.mu.Lock()
+	defer g.mu.Lock()
```
Comment (mrtracy) [inline gossip/gossip.go]: Defer Unlock()

File: gossip/gossip.go [hunk 2]
```diff
@@ -454,61 +454,81 @@ func (g *Gossip) manage(stopper *util.Stopper) {
...
-			g.mu.Unlock()
 		}
-	})
+	}
+	g.maybeSignalStalledLocked()
+}
+
+func (g *Gossip) doDisconnected(stopper *util.Stopper, c *client) {
+	g.mu.Lock()
+	defer g.mu.Lock()
```
Comment (mrtracy) [inline gossip/gossip.go]: Defer Unlock()

**Comment by tschottdorf** [comment]
LGTM

**Comment by mrtracy** [comment]
LGTM

**Comment by bdarnell** [comment]
LGTM

**Pull request merged**